### DISCUSSION TITLE
Nx 12 Roadmap

### DISCUSSION THREAD

**Discussion by vsavkin** [nrwl/nx#4645] [General]
# Nx 10 => Nx 11 Highlights

Read more about Nx 11 in this [blog post](https://blog.nrwl.io/introducing-nx-11-new-react-features-angular-11-incremental-builds-and-more-83c1e3a5210b).

Nx 11 is our biggest release yet, and I'm very proud of the core team working hard to make this happen. I’m also happy to see many folks in the community raising issues, submitting high quality PRs, creating Nx content.

Some highlights of the release:

## Core

* Introduced [Nx Devkit](https://nx.dev/latest/core-concepts/nx-devkit) for writing plugins, workspace generators, and workspace executors. Folks already authored a lot of plugins using util functions we provided, and using Angular Devkit. Nx Devkit should make this process simpler. More about the devkit later this month. 
* We are also going to reach out to folks authoring plugins to help them use the devkit and get their feedback.

## React

* Redesigned the home page of [nx.dev/react](https://nx.dev/react).
* Launched [a free Egghead.io course on Nx and React](https://egghead.io/playlists/scale-react-development-with-nx-4038).
* Published [a video comparing Nx and Lerna](https://www.youtube.com/watch?v=Edv7ZeI8Bwc).
* Added videos to [the Getting Started interactive tutorial](https://nx.dev/latest/react/tutorial/01-create-application).
* Cleaned up the ESLint setup for React.
* Improved support for incremental builds, which is useful for massive React apps.
* Improved Next.js support (thank you folks from Vercel for your help!)
* Introduced [React Native plugin](https://github.com/nrwl/nx-react-native)
* Introduced [Gatsby plugin](https://github.com/nrwl/gatsby)

## Node

* Added [the Node flavour of docs](https://nx.dev/node).
* Added [an interactive tutorial with videos.](https://nx.dev/latest/node/tutorial/01-create-application)
* Improved support for incremental builds.

## Angular

* Redesigned [nx.dev/angular](https://nx.dev/angular)
* Made ESLint the default linter when creating new workspaces with Angular. This reflects the quality of the ESLint plugin. Everyone creating new projects should be using it.
* Improved support for incremental builds, which is useful for massive Angular apps.

## Jest, Cypress, Storybook

Versions of Jest, Cypress, Storybook have been updated. 

## Nx Cloud Github Integration

Nx Cloud is a separate from Nx itself project that provides additional affordances for Nx workspaces (e.g., distributed caching). It now has free Github Integration. 

<img src="https://miro.medium.com/max/1400/0*8CNdDSQP9UDTHT4k">

If you have contributed to Nx, you might have seen this already. We think this is a great way to visualize runs, and it works a lot better than looking at, say, Circle logs directly. It also helps you troubleshoot cache misses and provides other insights.

To enable the GitHub integration install the following [Github App](https://github.com/apps/nx-cloud). To set up Nx Cloud go to [nx.app](https://nx.app). You can also run yourself using [this docker image](https://hub.docker.com/repository/docker/nxprivatecloud/nxcloud)).

## Repo Management & Backlog

Issues on github do several things:

- They allow folks to communicate with each other (troubleshooting, feature suggestions).
- They act as a backlog.
- They act as a feedback mechanism helping the core team focus on what is important.

These things can be in conflict. When we have too many issues (e.g., feature suggestions) that aren't up to date and haven't been updated in a long time, it's harder to communicate, and it's harder to know what is important. In other words, there is a tension between maintaining the backlog and being focused.

To help with this we have added a bot which closes stale issues. Closed issues are still searchable, they can be referenced, they can be reopened. This by no means implies that we care less about submitted issues. I personally read every single submitted issue. It just helps the core team and the community know what is still relevant.

## Updating to Nx 11

Nx comes with [a command to help you migrate your workspace to the new version.](https://nx.dev/latest/core-concepts/updating-nx). Check it out if you are migrating from Nx 10 to Nx 11.

<hr>

# Nx 12 Plans

Since Nx 11 got delayed, we only have 4 months before the Nx 12 release. As a result, the scope is a bit smaller.

The core team will focus on: bug fixes, refactoring, improved docs, improved support for plugins.

## Quality

We are starting the year by fixing bugs, paying technical debt, doing necessary refactorings. In addition, we are also adding extra checks to the CI pipeline to run the e2e tests on both Windows and Linux, and to test pnpm-based repos.

In general, we review the number of issues and discuss high-priority issues on a weekly basis, so the project doesn't go out of hand as many OSS projects do. This is **an additional effort** to make sure we start the year with a strong foundation.

## Plugin Support & Community Involvement

I'm delighted (and honestly a bit surprised) to see how many [Nx plugins](https://nx.dev/nx-community) folks have built. Thank you! Being the "VSCode of build tools" is what we always had in mind, and this is still our Number 1 priority.

**Devkit**

Nx Devkit is a simpler way to author plugins. We prioritize the ease of use and the ease of debugging. For instance, Nx Devkit only uses language primitives and simple utility functions. There is no monadic-style computation, no partial application etc.. We are migrating our generators and executors to Nx Devkit. This process should be wrapped up by mid-February.

For Angular folks who use the Angular devkit to write generators and builders/executors, we made using Angular Devkit from within Nx Devkit (and vice versa) seamless, so you should not be affected.

Finally, even though we have added the `@nrwl/devkit` package, the utility functions provided by `@nrwl/workspace` will remain there until at least Nx 14, and, depending on the feedback we will get, might remain there indefinitely.

**More Extension Points**

Almost anything in Nx can be extended. We are going to add two more extension points so the plugins can customize the calculation of the computation key and the code change analysis. This is mostly useful for non-JS plugins. 

**Involving Nx Community**

We are planning to launch a program for Nx community leaders (and in particular plugin authors). This is to recognize folks providing a lot of value, and to help organize the communication between the core team and the community.

## Better Support for Massive Workspaces

Many folks we work with have hundreds of projects in their workspaces. Cache computation and code change analysis already make Nx a great tool for this use case, but we can do better. For instance, currently the root configuration file can become a bit unruly when you have say 500 projects in your workspace. So we are going to support splitting it into multiple config files located in the projects' folders.

We are also working on refining our computation cache logic to reduce the number of times we invalidate all projects in the workspace.

Finally, even though we have made a lot of progress on making the [incremental build setup ergonomic](https://nx.dev/latest/guides/ci/incremental-builds#incremental-builds), there are more improvements coming, especially for React and Next.js folks.

## Nx.Dev

* We are updating the design to improve the discoverability of the content.
* We are redesigning the Node landing page to look similar to the Angular and React ones.
* We are updating the docs to clearly delineate between the introductory content and more advanced deep dive content.
* We are also planning to rework a lot of the content to make it more accessible for less experienced folks.
* We are adding more Node-related guides.

## React

* We are adding support for Detox test runner for React Native apps.
* We are updating the React builders to support Fast Refresh so code changes will be instantaneously applied without losing component state. New apps will have Fast Refresh enabled, while older apps can opt-in by updating their `workspace.json` file. 
* We are moving the Gatsby and React Native plugins into the main repo. As a result of that Gatsby and React Native docs will be available at [nx.dev/react](https://nx.dev/react]).

## Angular

* We are going to provide a schematic helping folks migrate existing workspaces from TSLint to ESLint.
* We are also planning to update the introductory video course for Angular devs.

**Comment by demisx** [comment]
Excellent work, guys. Our team top two wishes for NX 12:

1. Solid implementation of ability to auto-generate `package.json` per each app. As @emregency has eloquently put in  https://github.com/nrwl/nx/issues/1777 it should be **"centralized control, decentralized execution"**.
2. More docs on the best practices and recommended organization of libraries, directory structure, etc.

**Reply by virus2016** [reply]
Agreed! If we had this our teams would jump to NX in a heartbeat.

**Reply by beeman** [reply]
> More docs on the best practices and recommended organization of libraries, directory structure, etc.

@demisx with regards to your second point, I highly recommend reading [this free book](
https://go.nrwl.io/angular-enterprise-monorepo-patterns-new-book) by Nrwl. 

It's a a bit dated, and aimed towards Angular but the concepts stay the same and it should be fairly easy to map towards other technologies if that's what you use.

The book is pretty short and you can most likely finish it in a few hours. 

I also have a [few hours of live-streams](https://www.youtube.com/playlist?list=PL26p7XV-o0gsjLsqf59okAEWMRO5ApVkx) around structuring Nx in my YouTube channel, happy to share if you're interested.

**Reply by demisx** [reply]
Thank you. I've read that one already. It's a good start to understand the very basics, but it is Angular specific and didn't answer the questions we've run into when developing an enterprise level app. I remember having hard time finding best practices when developing back in 2019-2020. Especially, concerning backend NestJs development.

**Comment by maikdiepenbroek** [comment]
> Excellent work, guys. Our team top two wishes for NX 12:
> 
> 1. Solid implementation of ability to auto-generate `package.json` per each app. As @emregency has eloquently put in  #1777 it should be **"centralized control, decentralized execution"**.
> 2. More docs on the best practices and recommended organization of libraries, directory structure, etc.

Especially point 2 would be highly appreciated!

**Comment by cyrus-za** [comment]
Can we get #4482 in NX 13 roadmap please?

**Comment by prasanthkarukkuvel** [comment]
Support/Guidelines for Module Federation #3116 ?

**Comment by virus2016** [comment]
To add onto the wishes, it would be great to have versioning from something like semantic-release for each package. 

https://github.com/nrwl/nx/discussions/4645#discussioncomment-322027

**Reply by LayZeeDK** [reply]
Are you aware of [@jscutlery/semver](https://www.npmjs.com/package/@jscutlery/semver)?

**Comment by nasreddineskandrani** [comment]
Hi @vsavkin,
WCAG tools exist already in NX? or this need to be added?

Some informations about the tools I use personally if this can help this proposal.
1. On the pipeline

I suggest to look at https://github.com/dequelabs/axe-core (can be used with any website with the cli command)

2. if  you can **not** run the cli command onjenkins or cicleci. It can be done directly during dev:

**React:**
https://www.npmjs.com/package/@axe-core/react
also https://github.com/jsx-eslint/eslint-plugin-jsx-a11y
I don't know if this eslint plugin has much value on top of `@axe-core/react`

**Angular:**
codelyzer has something now.
https://medium.com/ngconf/new-accessibility-rules-in-codelyzer-v5-0-0-85eec1d3e9bb

**Comment by vitorcamachoo** [comment]
Any plans to support Yarn 2 with pnp mode active ?

**Reply by nemonemi** [reply]
I'd really like to see this supported as well.

**Comment by nealeu** [comment]
Good to see "Better Support for Massive Workspaces" in there.

I hope this is going to include some logic around finding nx.json and workspace.json based on the folder you are currently in.  If I'm down in apps/app/routes/, I should be able to run `nx build app` or even just `nx build` if there is a workspace.json in apps/app/

**Reply by thoaihuynh2509** [reply]
@nealeu  Can you share link about it? 
I cannot find it.

**Comment by DaveLo** [comment]
> Many folks we work with have hundreds of projects in their workspaces. Cache computation and code change analysis already make Nx a great tool for this use case, but we can do better. For instance, currently the root configuration file can become a bit unruly when you have say 500 projects in your workspace. So we are going to support splitting it into multiple config files located in the projects' folders.

We usually keep pretty close to the trunk, but having had my first rebase a stale branch experience with workspace and nx json files I can not express how happy this would make our team 😂

**Reply by violabg** [reply]
What happened to React builders to support Fast Refresh?
